File: android/app/src/main/kotlin/com/tampagamingguild/tgg_mobile/BeaconBackgroundChannel.kt
package com.tampagamingguild.tgg_mobile

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

/**
 * The control surface Dart uses to arm and disarm background beacon
 * detection, registered on the main engine (see MainActivity).
 *
 * The Dart entrypoint that should run on a detection is recorded here as a
 * raw callback handle while the app is alive, because a broadcast can arrive
 * long after the process is gone and the runner needs to know what to
 * execute (see BeaconBackgroundRunner).
 */
object BeaconBackgroundChannel {
    const val CONTROL_CHANNEL = "com.tampagamingguild.tgg_mobile/beacon_control"

    private const val PREFS_NAME = "tgg_beacon_background"
    private const val KEY_CALLBACK_HANDLE = "callback_handle"

    fun callbackHandle(context: Context): Long =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).getLong(KEY_CALLBACK_HANDLE, 0L)

    private fun setCallbackHandle(context: Context, handle: Long) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putLong(KEY_CALLBACK_HANDLE, handle)
            .apply()
    }

    fun register(context: Context, messenger: BinaryMessenger) {
        val appContext = context.applicationContext
        MethodChannel(messenger, CONTROL_CHANNEL).setMethodCallHandler { call, result ->
            when (call.method) {
                "registerCallback" -> {
                    val handle = call.argument<Any>("handle")
                    val raw = when (handle) {
                        is Long -> handle
                        is Int -> handle.toLong()
                        else -> null
                    }
                    if (raw == null) {
                        result.error("bad_args", "registerCallback needs a handle", null)
                    } else {
                        setCallbackHandle(appContext, raw)
                        result.success(null)
                    }
                }

                "startScan" -> {
                    val uuid = call.argument<String>("uuid")
                    if (uuid.isNullOrBlank()) {
                        result.error("bad_args", "startScan needs a uuid", null)
                    } else {
                        // Null means started; a string is a human-readable
                        // reason it couldn't, which Dart logs rather than
                        // surfaces, since this is all opportunistic.
                        result.success(BeaconScanControl.start(appContext, uuid))
                    }
                }

                "stopScan" -> {
                    BeaconScanControl.stop(appContext)
                    result.success(null)
                }

                else -> result.notImplemented()
            }
        }
    }
}

File: ios/Runner/BeaconBackgroundChannel.swift
import Flutter
import Foundation

/// Control surface Dart uses to arm/disarm background beacon detection --
/// iOS counterpart of BeaconBackgroundChannel.kt. The Dart callback handle
/// is stored in UserDefaults rather than SharedPreferences, since a region
/// entry can relaunch the process long after the app that registered the
/// callback is gone.
enum BeaconBackgroundChannel {
    // Must match _channel in beacon_background.dart exactly. These are
    // opaque names, so a mismatch compiles and launches cleanly and then
    // simply never delivers a call -- change both sides together.
    static let controlChannelName = "com.tampagamingguild.tggmobile/beacon_control"

    private static let callbackHandleKey = "beacon_background_callback_handle"

    /// 0 means unset, matching Android's SharedPreferences default of 0L.
    static var callbackHandle: Int64 {
        Int64(UserDefaults.standard.integer(forKey: callbackHandleKey))
    }

    static func register(with messenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(name: controlChannelName, binaryMessenger: messenger)
        channel.setMethodCallHandler { call, result in
            switch call.method {
            case "registerCallback":
                guard let args = call.arguments as? [String: Any],
                      let handle = args["handle"] as? NSNumber else {
                    result(FlutterError(code: "bad_args", message: "registerCallback needs a handle", details: nil))
                    return
                }
                UserDefaults.standard.set(handle.int64Value, forKey: callbackHandleKey)
                result(nil)

            case "startScan":
                guard let args = call.arguments as? [String: Any],
                      let uuid = args["uuid"] as? String, !uuid.isEmpty else {
                    result(FlutterError(code: "bad_args", message: "startScan needs a uuid", details: nil))
                    return
                }
                // nil means started; a string is a human-readable reason it
                // couldn't, which Dart logs rather than surfaces, since this
                // is all opportunistic.
                result(BeaconMonitor.shared.start(uuidString: uuid))

            case "stopScan":
                BeaconMonitor.shared.stop()
                result(nil)

            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
}
